Fix and improve raise_comms and sink_waits

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130573
* __->__ #129980

The tests for `raise_comms` and `sink_waits` passes were not enabled in CI. The passes are now broken due to functional collective v2 and possibly other changes.

Correctness issues:
- The original passes did not take mutation into consideration and may yield semantically different scheduling order. This may be due to the recent changes to how mutations are expressed in Inductor IR (e.g., MutationOutput).

Effectiveness issues:
- The original passes only moved the comm/wait nodes themselves. However, comm nodes can come with prologues (e.g., clone for all_reduce_, split-cat for non-zero dim all-gather). Whenever there are any prologues, the comms won't be raised at all.
- The prologues are often horizontally fused with other pointwise nodes. This can severely delay the scheduling of the comm node.

This PR:
- Make the passes handle mutation correctly.
- Instead of moving individual comm/wait nodes, schedule all node using a scored method. This way the comm nodes can be optimally raised even in the presence of prologues.
- The horizontal fusion of prolofues often severely delays the scheduling of the comm node. Horizontally fusing this clone can almost never out-perform scheduling the comm node earlier. Also in most cases, this clone is eliminated via in-place reuse. Therefore, we tell the scheduler to not fuse it.
- Enable the tests in CI.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @tianyu-l

Commit:
Fix and improve raise_comms and sink_waits

[ghstack-poisoned]

Chillee commented on torch/_inductor/comms.py:
```diff
@@ -17,61 +20,116 @@
+
+    unmet_deps: Dict[BaseSchedulerNode, Set[str]] = {}
+    for snode in snodes:
+        # A mutating node's unmet_dependencies doesn't cover the dependencies
```
I don't think this should be needed - the mutation dependencies should already be encoded in the scheduler node's edges.

yifuwang replied:
I thought so too but it doesn't seem to be the case (learned it the hard way). For example, we have:
```
node = (...)
mutation_output = (node_doing_mutating=node, ...)
```

`mutation_output` would have more `unmet_dependencies` than `node`.

Maybe we should fix it in the scheduler?

peterbell10 replied:
Yes `MutationOutput` expressed the dependencies incorrectly. This is fixed in #129346 though (the entire stack).

yifuwang replied:
@peterbell10 appreciate the fix!

yf225 replied:
after the rebase, we will need some changes (https://github.com/pytorch/pytorch/pull/130276) to make this algo working again (please feel free to merge my PR into this PR 🙂)

yf225 commented on test/distributed/test_compute_comm_reordering.py:
```diff
@@ -146,25 +164,43 @@ def func(a, *, tag, ranks, group_size):
+            # - The clone prologue of the all_reduce_ should not be fused with
+            # any relus.
+            # - The all_reduce_ and its prologue should be raised above the 2nd
+            # matmul but below the 1st matmul.
```
Loved the ability to raise prologue!

I wonder should we add another unit test / enhance this unit test to show it can raise multiple prologue ops? e.g. 
```
b = torch.matmul(a, a)
c = torch.relu(b)
d = torch.matmul(c, c)
b1 = b + b
e = _functional_collectives.all_reduce(b1, "sum", "0")
```
the (`b1`, `e`) computation block should raise above the (`c`, `d`) computation block. 

Thanks!

yf225 replied:
Added this enhancement in https://github.com/pytorch/pytorch/pull/130276/files#diff-5711354a54dd404142d5444e489df93d4a303bdc74b302ef64dfbb6d68b88e6aR170

yf225 commented on torch/_inductor/comms.py:
```diff
@@ -17,61 +20,116 @@
+    name_to_snode = {}
+    scores_0, scores_1, scores_2 = {}, {}, {}
+    for idx, snode in enumerate(snodes):
+        for name in snode.get_names():
```
To merge to latest master, we might need:
```suggestion
        for name in snode.get_buffer_names():
```

yf225 commented on torch/_inductor/comms.py:
```diff
@@ -17,61 +20,116 @@
+    class Runnable:
+        def __init__(self, snode):
+            self.snode = snode
+            name = next(iter(snode.get_names()))
```
```suggestion
            name = next(iter(snode.get_buffer_names()))
```

yf225 commented on torch/_inductor/comms.py:
```diff
@@ -17,61 +20,116 @@
+    while len(ready):
+        curr = heapq.heappop(ready).snode
+        scheduled.append(curr)
+        for curr_name in curr.get_names():
```
```suggestion
        for curr_name in curr.get_buffer_names():
```

yf225 commented on torch/_inductor/scheduler.py:
```diff
@@ -2276,6 +2276,10 @@ def can_fuse(self, node1: BaseSchedulerNode, node2: BaseSchedulerNode) -> bool:
             why("template epilogue not satisfied")
             return False
 
+        if (node1.get_names() | node2.get_names()) & V.graph.no_fuse_buffer_names:
```
```suggestion
        if (node1.get_buffer_names() | node2.get_buffer_names()) & V.graph.no_fuse_buffer_names:
```

malfet reviewed (changes requested):
You don't need to update neither CuDNN nor Kineto, do you?

yifuwang commented:
@malfet yup that was a mistake

yf225 committed:
Update

[ghstack-poisoned]

yf225 commented on torch/_inductor/comms.py:
```diff
@@ -2,6 +2,9 @@
 # pyre-strict
```
might need to remove the `third_party/cudnn_frontend` changes

Commit:
Update on "Fix and improve raise_comms and sink_waits"


The tests for `raise_comms` and `sink_waits` passes were not enabled in CI. The passes are now broken due to functional collective v2 and possibly other changes.

Correctness issues:
- The original passes did not take mutation into consideration and may yield semantically different scheduling order. This may be due to the recent changes to how mutations are expressed in Inductor IR (e.g., MutationOutput).

Effectiveness issues:
- The original passes only moved the comm/wait nodes themselves. However, comm nodes can come with prologues (e.g., clone for all_reduce_, split-cat for non-zero dim all-gather). Whenever there are any prologues, the comms won't be raised at all.
- The prologues are often horizontally fused with other pointwise nodes. This can severely delay the scheduling of the comm node.

This PR:
- Make the passes handle mutation correctly.
- Instead of moving individual comm/wait nodes, schedule all node using a scored method. This way the comm nodes can be optimally raised even in the presence of prologues.
- The horizontal fusion of prolofues often severely delays the scheduling of the comm node. Horizontally fusing this clone can almost never out-perform scheduling the comm node earlier. Also in most cases, this clone is eliminated via in-place reuse. Therefore, we tell the scheduler to not fuse it.
- Enable the tests in CI.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab chauhang d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire

[ghstack-poisoned]

yifuwang commented:
Squashed fixes from @yf225 (https://github.com/pytorch/pytorch/pull/130276)

yifuwang commented:
@pytorchbot merge

Commit:
Update on "Fix and improve raise_comms and sink_waits"


The tests for `raise_comms` and `sink_waits` passes were not enabled in CI. The passes are now broken due to functional collective v2 and possibly other changes.

Correctness issues:
- The original passes did not take mutation into consideration and may yield semantically different scheduling order. This may be due to the recent changes to how mutations are expressed in Inductor IR (e.g., MutationOutput).

Effectiveness issues:
- The original passes only moved the comm/wait nodes themselves. However, comm nodes can come with prologues (e.g., clone for all_reduce_, split-cat for non-zero dim all-gather). Whenever there are any prologues, the comms won't be raised at all.
- The prologues are often horizontally fused with other pointwise nodes. This can severely delay the scheduling of the comm node.

This PR:
- Make the passes handle mutation correctly.
- Instead of moving individual comm/wait nodes, schedule all node using a scored method. This way the comm nodes can be optimally raised even in the presence of prologues.
- The horizontal fusion of prolofues often severely delays the scheduling of the comm node. Horizontally fusing this clone can almost never out-perform scheduling the comm node earlier. Also in most cases, this clone is eliminated via in-place reuse. Therefore, we tell the scheduler to not fuse it.
- Enable the tests in CI.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab chauhang d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire

[ghstack-poisoned]

Commit:
Update on "Fix and improve raise_comms and sink_waits"


The tests for `raise_comms` and `sink_waits` passes were not enabled in CI. The passes are now broken due to functional collective v2 and possibly other changes.

Correctness issues:
- The original passes did not take mutation into consideration and may yield semantically different scheduling order. This may be due to the recent changes to how mutations are expressed in Inductor IR (e.g., MutationOutput).

Effectiveness issues:
- The original passes only moved the comm/wait nodes themselves. However, comm nodes can come with prologues (e.g., clone for all_reduce_, split-cat for non-zero dim all-gather). Whenever there are any prologues, the comms won't be raised at all.
- The prologues are often horizontally fused with other pointwise nodes. This can severely delay the scheduling of the comm node.

This PR:
- Make the passes handle mutation correctly.
- Instead of moving individual comm/wait nodes, schedule all node using a scored method. This way the comm nodes can be optimally raised even in the presence of prologues.
- The horizontal fusion of prolofues often severely delays the scheduling of the comm node. Horizontally fusing this clone can almost never out-perform scheduling the comm node earlier. Also in most cases, this clone is eliminated via in-place reuse. Therefore, we tell the scheduler to not fuse it.
- Enable the tests in CI.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab chauhang d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire

[ghstack-poisoned]

Commit:
Update on "Fix and improve raise_comms and sink_waits"


The tests for `raise_comms` and `sink_waits` passes were not enabled in CI. The passes are now broken due to functional collective v2 and possibly other changes.

Correctness issues:
- The original passes did not take mutation into consideration and may yield semantically different scheduling order. This may be due to the recent changes to how mutations are expressed in Inductor IR (e.g., MutationOutput).

Effectiveness issues:
- The original passes only moved the comm/wait nodes themselves. However, comm nodes can come with prologues (e.g., clone for all_reduce_, split-cat for non-zero dim all-gather). Whenever there are any prologues, the comms won't be raised at all.
- The prologues are often horizontally fused with other pointwise nodes. This can severely delay the scheduling of the comm node.

This PR:
- Make the passes handle mutation correctly.
- Instead of moving individual comm/wait nodes, schedule all node using a scored method. This way the comm nodes can be optimally raised even in the presence of prologues.
- The horizontal fusion of prolofues often severely delays the scheduling of the comm node. Horizontally fusing this clone can almost never out-perform scheduling the comm node earlier. Also in most cases, this clone is eliminated via in-place reuse. Therefore, we tell the scheduler to not fuse it.
- Enable the tests in CI.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab chauhang d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge

Chillee commented on torch/_inductor/comms.py:
```diff
@@ -17,61 +20,124 @@
+    # We can remove this logic once the fix is landed.
+    unmet_deps: Dict[BaseSchedulerNode, Set[str]] = {}
+    for snode in snodes:
+        if isinstance(snode.node, ir.MutationOutput):
```
Do we still need this change after @peterbell10's fix?

yifuwang replied:
No we can remove it after the fixes land. @yf225 and I decided to land this version first since we both have changes that depend on this PR (also this PR fixes the tests and enables them in CI).
